ui/src/types/queries.ts

```diff
@@ -1,10 +1,43 @@
+import {CUSTOM_TIME_RANGE_LABEL} from 'src/shared/constants/timeRanges'
+
 export {Query, Dialect} from 'src/client'
 
-export interface TimeRange {
-  lower: string
-  upper?: string | null
-  seconds?: number
+export type SelectableTimeRangeLower =
+  | 'now() - 5m'
+  | 'now() - 15m'
+  | 'now() - 1h'
+  | 'now() - 6h'
+  | 'now() - 12h'
+  | 'now() - 24h'
+  | 'now() - 2d'
+  | 'now() - 7d'
+  | 'now() - 30d'
+
+export type TimeRange =
+  | SelectableDurationTimeRange
+  | DurationTimeRange
+  | CustomTimeRange
+
+export interface SelectableDurationTimeRange {
+  lower: SelectableTimeRangeLower
+  upper: null
+  seconds: number
   format?: string
-  label?: string
-  duration?: string
+  label: string
+  duration: string
+  type: 'selectable-duration'
+}
+
+export interface DurationTimeRange {
+  lower: string
+  upper: null
+  label: string
+  type: 'duration'
+}
+
+export interface CustomTimeRange {
+  lower: string
+  upper: string
+  label: typeof CUSTOM_TIME_RANGE_LABEL | string
+  type: 'custom'
```

Comment:
This is great, huge improvement from what's there, which is all willy nilly.

I've been marinating on this for a couple hours and I want to get your thoughts on a tweak to the types to see if it's a little more expressive.

So there are a couple of concepts I'm seeing here (I'm naming them after their high level functionality, the names may not map to the variable names in this pr):

* `TimeRange` (as a concept, not as a construct in our application): this is two numbers, a lower and an upper. More formally, this is [an ordered pair](https://en.wikipedia.org/wiki/Ordered_pair) or a [2-tuple](https://www.typescriptlang.org/docs/handbook/basic-types.html#tuple).
* `AppTimeRange`: this has a `TimeRange` (a lower and an upper), a `label`, and a `type`. `Duration` and `Custom` have an `is a` relationship with `AppTimeRange` and `has a` relationship with `TimeRange`
  * A more specific type of `AppTimeRange` is a `SelectableDurationTimeRange`, which is defined by a select dropdown, has no `upper`, and has a specific set of values allowed for `lower`.

Does that jive with what you've seen and are expressing? With all that above said, what do you think of this tweak:

```ts
export type SelectableTimeRangeLower =
  | 'now() - 5m'
  | 'now() - 15m'
  | 'now() - 1h'
  | 'now() - 6h'
  | 'now() - 12h'
  | 'now() - 24h'
  | 'now() - 2d'
  | 'now() - 7d'
  | 'now() - 30d'

export type TimeRange = [string, string | null]

// if it's easier to make TimeRange an interface cause of upper and lower existing all over:
export interface TimeRange {
  lower: string
  upper: string | null
}

export type DashboardRangeType = 'custom' | 'duration' | 'selectable-duration'

// this name may or may not suck, please don't be attached to it if it does
export interface DashboardRange {
  label: string
  range: TimeRange
  type: DashboardRangeType
}

export interface SelectableDashboardRange extends DashboardRange {
  duration: string
  format?: string
  range: [SelectableTimeRangeLower, null]
  seconds: number
}
```
Here's what I'm thinking

1. This makes `TimeRange` a little more reusable and composable, and I think expressive - a range is two date strings - a 2-tuple expresses that very well
   1. I saw a place or two in this PR where a function needed a range-like structure - that kind of inspired the idea that this should be its own concept cause it seems applicable
1. This makes use of [Liskov Substitution](https://en.wikipedia.org/wiki/Liskov_substitution_principle) which is really slick when it works (I think this is one of those cases where it works really well). Basically anywhere you need a dashboard range, you type hint on `DashboardRange`, and any `DashboardRange` or `SelectableDashboardRange` (which extends `DashboardRange`) will work. Any more spefic range we make in the future based off `DashboardRange` will work
1. If this needs to be more specific (for example you want to only allow specific values for `label` on `CustomTimeRange`, we could just do something as simple as:
    ```ts
    export interface CustomTimeRange extends DashboardRange {
      range: [string, string]
      label: typeof CUSTOM_TIME_RANGE_LABEL | string
    }
    ```
So what do you think of that? I'm available to chat on a video call if that's too much words!

Reply:
I will remove `label`